hub/api/pytorch.py

```diff
@@ -0,0 +1,241 @@
+
+
+def shared_memory_clear(chunk_set):
+    """Helper function that checks if an existing SharedMemory exists for any chunk in chunk_set and clears it"""
+    for chunk_name in chunk_set:
+        try:
+            shm = SharedMemory(name=chunk_name)
+            shm.close()
+            shm.unlink()
+        except:
+            pass
+
+
+def read_chunk(chunk, key):
+    """Reads a single chunk from the dataset's storage provider and stores it in the SharedMemory"""
+    remove_shared_memory_from_resource_tracker()
+    chunk_path = os.path.join(key, "chunks", chunk)
+    chunk_bytes = _hub_storage_provider[chunk_path]
+    chunk = chunk_path.split("/")[-1]
+    shm = SharedMemory(create=True, size=len(chunk_bytes), name=chunk)
+    shm.buf[:] = chunk_bytes
+    shm.close()
+    return
+
+
+def remove_shared_memory_from_resource_tracker():
+    """Helper function to fix bug in Python SharedMemory
+
+    Monkey-patch multiprocessing.resource_tracker so SharedMemory won't be tracked
+    More details at: https://bugs.python.org/issue38119
+    """
+
+    def fix_register(name, rtype):
+        if rtype == "shared_memory":
+            return
+        return resource_tracker._resource_tracker.register(self, name, rtype)
+
+    def fix_unregister(name, rtype):
+        if rtype == "shared_memory":
+            return
+        return resource_tracker._resource_tracker.unregister(self, name, rtype)
+
+    resource_tracker.register = fix_register
+    resource_tracker.unregister = fix_unregister
+    if "shared_memory" in resource_tracker._CLEANUP_FUNCS:
+        del resource_tracker._CLEANUP_FUNCS["shared_memory"]
+
+
+def _to_pytorch(dataset, transform=None, workers=1):
+    try:
+        import torch
+    except ModuleNotFoundError:
+        raise ModuleNotInstalledException("torch")
+    global torch
+    global _hub_storage_provider  # global to pass to processes, as not possible to serialize it and send as argument
+
+    # TODO, apparently boto3.client isn't safe for multiprocessing https://github.com/boto/boto3/pull/2848/files
+    # could it be working here as we're only reading data?
+    _hub_storage_provider = dataset.provider
+    return TorchDataset(dataset, transform, workers)
+
+
+class TorchDataset:
+    def __init__(self, dataset, transform=None, workers=1):
+        self.dataset = dataset  # TODO disable the memory cache
+        self.transform = transform
+        self.workers = workers
+        self.storage = self.dataset.provider
+        self.map = ProcessPool(nodes=workers).map
+
+        # contains index_map for each Tensor
+        self.all_index_maps = self.load_index_maps()
+
+        # contains meta for each Tensor
+        self.all_meta = self.load_meta()
+
+        # corresponding to each Tensor, stores index-value map. here value is the actual array at the index for the key
+        # this essentially acts as our in memory prefetch cache
+        self.all_index_value_maps = defaultdict(dict)
```

Comment:
is a good variable name for this `self.prefetch_cache`? this long comment is kind of confusing